### PR TITLE
MULTIARCH-4550: Power VS: Use a custom resolver when publish strategy is Internal

### PR DESCRIPTION
**Pull request by mjturek**

### CODE REVIEW

File: pkg/asset/installconfig/powervs/client.go [hunk 1]
```diff
@@ -510,6 +555,24 @@ func (c *Client) GetPublicGatewayByVPC(ctx context.Context, vpcName string) (*vp
 	return nil, nil
 }
 
+// GetVPCSubnets retrieves all subnets in the given VPC.
+func (c *Client) GetVPCSubnets(ctx context.Context, vpcID string) ([]vpcv1.Subnet, error) {
+	_, cancel := context.WithTimeout(ctx, 1*time.Minute)
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:560]: ditto

File: pkg/asset/installconfig/powervs/client.go [hunk 2]
```diff
@@ -268,6 +271,34 @@ func (c *Client) GetInstanceCRNByName(ctx context.Context, name string, publish
 	return "", fmt.Errorf("DNS zone %q not found", name)
 }
 
+// GetDNSCustomResolverIP gets the DNS Server IP of a custom resolver associated with the specified VPC subnet in the specified DNS zone.
+func (c *Client) GetDNSCustomResolverIP(ctx context.Context, dnsID string, vpcID string) (string, error) {
+	_, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:277]: You're creating a child context but discarding it and passing the parent context to the calls inside this function. I'm guessing that was not your intention? From what I see below, the calls you pass the context to create their own child contexts, so you don't really need this.
```suggestion

```

File: pkg/asset/installconfig/powervs/client.go [hunk 3]
```diff
@@ -390,6 +421,20 @@ func (c *Client) GetDNSInstancePermittedNetworks(ctx context.Context, dnsID stri
 	return networks, nil
 }
 
+// CreateDNSCustomResolver creates a custom resolver for the VPC.
+func (c *Client) CreateDNSCustomResolver(ctx context.Context, dnsID string, dnsZone string) (*dnssvcsv1.CustomResolver, error) {
+	_, cancel := context.WithTimeout(ctx, 1*time.Minute)
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:426]: Do not discard the child context returned.
```suggestion
	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
```

File: pkg/asset/installconfig/powervs/client.go [hunk 4]
```diff
@@ -510,6 +555,24 @@ func (c *Client) GetPublicGatewayByVPC(ctx context.Context, vpcName string) (*vp
...
+	listSubnetsOptions.VPCID = &vpcID
+	subnets, _, err := c.vpcAPI.ListSubnets(listSubnetsOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []vpcv1.Subnet{}
+	result = append(result, subnets.Subnets...)
+
+	return result, nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:573]:
```suggestion
	return subnets.Subnets, nil
```

File: pkg/asset/installconfig/powervs/metadata.go [hunk 5]
```diff
@@ -240,3 +241,40 @@ func (m *Metadata) IsVPCPermittedNetwork(ctx context.Context, vpcName string, ba
...
+}
+
+// GetVPCSubnets gets a list of subnets in a VPC.
+func (m *Metadata) GetVPCSubnets(ctx context.Context, vpcName string) ([]vpcv1.Subnet, error) {
+	vpc, err := m.client.GetVPCByName(ctx, vpcName)
+	if err != nil {
+		return nil, err
+	}
+	subnets, err := m.client.GetVPCSubnets(ctx, *vpc.ID)
+	return subnets, err
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/metadata.go:261]:
```suggestion
	return m.client.GetVPCSubnets(ctx, *vpc.ID)
```

File: pkg/asset/installconfig/powervs/client.go (hunk 2 above)
Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go:277]: Yep makes sense, bad copy paste

Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go:277]: Removed other occurrences of it as well. Hope that's okay.

File: pkg/asset/installconfig/powervs/client.go (hunk 3 above)
Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go:426]: Actually removing this whole function as it's not needed currently.

File: pkg/asset/installconfig/powervs/client.go (hunk 4 above)
Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go:573]: thanks!

File: pkg/asset/installconfig/powervs/metadata.go (hunk 5 above)
Comment (mjturek) [reply pkg/asset/installconfig/powervs/metadata.go:261]: Forgot to check if `err != nil` etc. Added that instead!

File: pkg/asset/installconfig/powervs/client.go (hunk 1 above)
Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go]: Switched to ListSubnetsWithContext and passed along `ctx` instead of creating a child one.

**Comment by mjturek** [comment]
Thanks for the review @r4f4 ! I think I addressed everything... Let me know what you think now

File: pkg/asset/installconfig/powervs/client.go [hunk 6]
```diff
@@ -268,6 +270,32 @@ func (c *Client) GetInstanceCRNByName(ctx context.Context, name string, publish
...
+	for _, customResolver := range customResolvers.CustomResolvers {
+		for _, location := range customResolver.Locations {
+			for _, subnet := range subnets {
+				if *subnet.CRN == *location.SubnetCrn {
+					return *location.DnsServerIp, nil
+				}
+			}
+		}
+	}
+	return "", nil
```
Comment (r4f4) [inline pkg/asset/installconfig/powervs/client.go:296]: Maybe it's better to return an error here? When using this function you seem to assume if `err == nil` than the IP returned is valid.
```suggestion
	return "", fmt.Errorf("DNS server IP of custom resolver for %q not found", dnsID)
```

Comment (mjturek) [reply pkg/asset/installconfig/powervs/client.go:294]: yep good point, committed!

**Pull request merged**